File: base_985161.js
// Base-985161 encoder/decoder (ES module + simple CLI)
// Radix: 985,161 (not a power of two)
// Alphabet: U+10000 .. U+10000+985160 (avoids surrogates)
// Method: big-integer base conversion (256 <-> 985,161) with leading-zero preservation

export const BASE = 985161;
export const CP_START = 0x10000;
export const CP_END   = CP_START + BASE - 1;

export function valToChar(v){
  if (v < 0 || v >= BASE) throw new RangeError('digit out of range');
  return String.fromCodePoint(CP_START + v);
}
export function charToVal(ch){
  const cp = ch.codePointAt(0);
  if (cp < CP_START || cp > CP_END) throw new Error(`Invalid Base-985161 character U+${cp.toString(16).toUpperCase()}`);
  return cp - CP_START;
}

// Divide big-int represented as digits[] (baseFrom, MSB-first) by divisor; return [quotientDigits, remainder]
function divmodNumber(digits, baseFrom, divisor){
  const q = [];
  let rem = 0;
  for (let i=0;i<digits.length;i++){
    const acc = rem * baseFrom + digits[i];
    const d = Math.floor(acc / divisor);
    rem = acc % divisor;
    if (q.length || d !== 0) q.push(d);
  }
  return [q, rem];
}

export function encode(bytes){
  if (!bytes || bytes.length === 0) return valToChar(0);

  // Count leading zero bytes
  let zeros = 0; while (zeros < bytes.length && bytes[zeros] === 0) zeros++;

  let digits256 = Array.from(bytes); // MSB-first
  const outVals = [];
  while (digits256.length){
    const [q, rem] = divmodNumber(digits256, 256, BASE);
    outVals.push(rem);
    digits256 = q;
  }
  // Add markers for leading zero bytes
  for (let i=0;i<zeros;i++) outVals.push(0);
  outVals.reverse();
  return outVals.map(valToChar).join('');
}

export function decode(text){
  if (!text || text.length === 0) return new Uint8Array();
  const digitsB = [];
  for (let i=0;i<text.length;){
    const cp = text.codePointAt(i);
    i += (cp > 0xFFFF ? 2 : 1);
    if (cp < CP_START || cp > CP_END) throw new Error(`Invalid Base-985161 digit U+${cp.toString(16).toUpperCase()}`);
    digitsB.push(cp - CP_START);
  }
  // Count leading zero-digits
  let zeros = 0; for (const v of digitsB){ if (v===0) zeros++; else break; }

  const out = [];
  let arr = digitsB.slice();
  while (arr.length){
    const [q, rem] = divmodNumber(arr, BASE, 256);
    out.push(rem);
    arr = q;
  }
  for (let i=0;i<zeros;i++) out.push(0);
  out.reverse();
  return new Uint8Array(out);
}

// -------------- CLI (Node) --------------
if (typeof process !== 'undefined' && import.meta.url === `file://${process.argv[1]}`){
  const fs = await import('node:fs');
  const args = process.argv.slice(2);
  if (args.length < 1 || !['enc','dec'].includes(args[0])){
    console.error('Usage: node base985161.js enc|dec [infile|-] [outfile|-]');
    process.exit(2);
  }
  const mode = args[0];
  const inf  = args[1] ?? '-';
  const outf = args[2] ?? '-';

  if (mode === 'enc'){
    const data = inf === '-' ? await readStdinBin() : fs.readFileSync(inf);
    const text = encode(new Uint8Array(data));
    if (outf === '-') process.stdout.write(text);
    else fs.writeFileSync(outf, text, 'utf8');
  } else {
    const text = inf === '-' ? await readStdinTxt() : fs.readFileSync(inf, 'utf8');
    const bytes = decode(text);
    if (outf === '-') process.stdout.write(Buffer.from(bytes));
    else fs.writeFileSync(outf, Buffer.from(bytes));
  }
}

async function readStdinBin(){ const chunks=[]; for await (const c of process.stdin) chunks.push(c); return Buffer.concat(chunks); }
async function readStdinTxt(){ const buf = await readStdinBin(); return buf.toString('utf8'); }

File: base985161.py
#!/usr/bin/env python3
"""
Base-985161 encoder/decoder (Python)

- Radix: 985,161 (not a power of two)
- Alphabet: contiguous Unicode code points U+10000 .. U+10000+985160
  (stays clear of surrogates; safe scalar values)
- Method: big-integer base conversion (256 <-> 985,161) with leading-zero preservation

Usage:
    python base985161.py enc input.bin output.b985161
    python base985161.py dec input.b985161 output.bin

    # or via stdin/stdout
    cat file | python base985161.py enc - - > out.b985161
    cat out.b985161 | python base985161.py dec - - > file
"""
from __future__ import annotations
import sys
from typing import List, Tuple

BASE = 985_161
CP_START = 0x10000
CP_END   = CP_START + BASE - 1

# ---------------- Alphabet mapping ----------------

def val_to_char(v: int) -> str:
    if not (0 <= v < BASE):
        raise ValueError("digit out of range")
    return chr(CP_START + v)


def char_to_val(ch: str) -> int:
    cp = ord(ch)
    if not (CP_START <= cp <= CP_END):
        raise ValueError(f"Invalid Base-985161 character U+{cp:04X}")
    return cp - CP_START

# ---------------- Core big-int helpers ----------------

def _divmod_number(digits: List[int], base_from: int, divisor: int) -> Tuple[List[int], int]:
    """Divide big-int given as digits (base_from, MSB-first) by `divisor`.
    Returns (quotient_digits (MSB-first, no leading zeros), remainder).
    """
    q: List[int] = []
    rem = 0
    for d in digits:
        acc = rem * base_from + d
        q_digit = acc // divisor
        rem = acc % divisor
        if q or q_digit != 0:
            q.append(q_digit)
    return q, rem

# ---------------- Public API ----------------

def encode(data: bytes) -> str:
    """Encode arbitrary bytes to Base-985161 text.
    Preserves leading 0x00 bytes by emitting leading zero-digits.
    """
    if not data:
        return ""

    # Count leading zero bytes
    zeros = 0
    for b in data:
        if b == 0:
            zeros += 1
        else:
            break

    # Convert base-256 digits -> base-985161 by repeated div/mod
    # Leading zeros are markers, not part of the numeric payload.
    digits256 = list(data[zeros:])  # MSB-first
    out_vals: List[int] = []
    while digits256:
        digits256, rem = _divmod_number(digits256, 256, BASE)
        out_vals.append(rem)
    # Most-significant first
    out_vals.reverse()
    return ''.join(val_to_char(v) for v in ([0] * zeros + out_vals))


def decode(text: str) -> bytes:
    """Decode Base-985161 text back to bytes.
    Restores leading zero bytes from leading zero-digits.
    """
    if not text:
        return b""

    # Convert text to base-BASE digits (MSB-first)
    digitsB = [char_to_val(ch) for ch in text]

    # Count leading zeros (digit value 0)
    zeros = 0
    for v in digitsB:
        if v == 0:
            zeros += 1
        else:
            break

    # Convert to base-256 via repeated division
    out: List[int] = []
    # Exclude zero markers from the numeric conversion.
    digitsB = digitsB[zeros:]
    while digitsB:
        digitsB, rem = _divmod_number(digitsB, BASE, 256)
        out.append(rem)
    # MSB-first: reverse remainders
    out.reverse()
    return bytes([0] * zeros + out)

# ---------------- CLI ----------------

def main(argv: List[str]) -> int:
    if len(argv) < 2 or argv[1] not in ("enc", "dec"):
        print("Usage: base985161.py enc|dec [infile|-] [outfile|-]", file=sys.stderr)
        return 2
    mode = argv[1]
    inf  = argv[2] if len(argv) > 2 else "-"
    outf = argv[3] if len(argv) > 3 else "-"

    if mode == "enc":
        if inf == "-":
            data = sys.stdin.buffer.read()
        else:
            with open(inf, "rb") as input_file:
                data = input_file.read()
        txt = encode(data)
        if outf == "-":
            sys.stdout.write(txt)
        else:
            with open(outf, "w", encoding="utf-8") as output_file:
                output_file.write(txt)
    else:
        if inf == "-":
            txt = sys.stdin.read()
        else:
            with open(inf, "r", encoding="utf-8") as input_file:
                txt = input_file.read()
        raw = decode(txt)
        if outf == "-":
            sys.stdout.buffer.write(raw)
        else:
            with open(outf, "wb") as output_file:
                output_file.write(raw)
    return 0

if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
